### ISSUE TITLE
Consider using shellcheck + linter in your ide

### ISSUE THREAD

**Issue by edge226** [manjaro/manjaro-tools#213]
I use Atom editor. It has a neat plugin system that allows linter + shellcheck to automatically lint your code.

Running `find ./ -type f -name '*.sh' -exec shellcheck {} + > shellcheck` I was able to produce this file with all of your errors + warnings mixed in together.

If you have shellcheck installed then it shows you red lines for errors and orange lines for warnings.

I've attached a file with all of the warnings and errors based on your scripts.

[shellcheck.txt](https://github.com/manjaro/manjaro-tools/files/257110/shellcheck.txt)

**Comment by edge226** [comment]
This really is not as difficult a fix as it seems @udeved . I personally use atom and I just install shellcheck + two atom plugins to have live linting as I type, linter and linter-shellcheck do this.

I've also got minimap and minimap is able to display the warning/error lines and what they are, This is done via another plugin which I believe is minimap-linter or something.

In the screen shot the bottom section takes me to that particular line with the error. I clicked the top one and it shows me the error. In this case it wants each variable declaration on its own line. A simple 2 second fix... Most of them are.
![util-iso sh -home-edge-github-manjaro-tools atom_091](https://cloud.githubusercontent.com/assets/4232080/15272150/5adbc24a-1a29-11e6-89b0-41d7779c3bb2.png)

**Comment by udeved** [comment]
I guess, it is entirely up to development to decide when and what feature is to be implemented.
Sorry, target will remain 1.0.0, period.

**Comment by edge226** [comment]
Of course its up to you when features are implemented. If you want to keep implementing new features on buggy code that may produce incorrect results that is entirely up to you.

Good luck with that development strategy.

**Comment by udeved** [comment]
I appreciate your report.
Hence I took it serious and assigned it to 1.0.0 milestone.
There is other stuff to sort out, and I am simply saying, you won't dictate development which I do in my spare time. I t will be tackled, but not now or in next milestone.

**Comment by edge226** [comment]
I'm not dictating your development. I'm stating the work is a lot less than you think it is. Its probably about an hour or two to look over to fix **every** bug in **all** of your files. I just do not see a point in leave bugs that are going to be hidden caveats in code there if they are extremely easy to fix.

I personally fix everything in my code immediately on simple premise of "Code talks", If you write code with hidden caveats and gotcha's. It reflects upon the author. I prefer what I write to reflect positively upon me, If you do not that is fine, Like I said it does not affect me.

While yes this is development you do in your spare time you're attempting to provide it in a production environment in a project that states it is community based, Yet not a lot of the code by the project is really created to be easy for people within the community to learn.

Personally its something I've always found strange about Manjaro. Caring so strongly whether things are user friendly or not... But not putting that care into the code.

/me shrugs.

**Comment by udeved** [comment]
If its all quickly done, send a pull request, piece of cake.
I don't have time for this right now, and lots of warning you posted are just that, warnings, comparable to compiler warnings.
If you can point out a grave and big bug do so. 
I do think you exaggerate a bit here.

**Comment by edge226** [comment]
Syntax warnings in a bash environment are a little different. Why did'nt I Just fix it and send a Pull Request?

What do you learn about your code and coding correctly then?

You think thats helping you but honestly it **is** doing you a dis-service if I fix your code.

Things like `^-- SC2086: Double quote to prevent globbing and word splitting.` are actually really dangerous because if someone say adds a * to something you do not expect it could overrun and do things in other directories. The word splitting thing means if you're running through entries it could split your entries in half and your program produces incorrect output.

`^-- SC2154: pkgver_override is referenced but not assigned.`
and
`^-- SC2034: pkgparts appears unused. Verify it or export it.`
Are absolutely harmless and due to sourcing another script. In atom you can actually stick these on ignore.

`^-- SC2128: Expanding an array without an index only gives the first element.`
This will output entirely incorrect information and if you wanted the first element you should be doing "${array[0]}"

^-- SC2005: Useless echo? Instead of 'echo $(cmd)', just use 'cmd'.
This one just damn well makes me laugh. Its like a damn joke.

^-- SC2091: Remove surrounding $() to avoid executing output.
That will literally cause your script to break by executing the if on the ouput of the command rather than checking the exit status of the command.

^-- SC2010: Don't use ls | grep. Use a glob or a for loop with a condition to allow non-alphanumeric filenames.
A user can probably overrun your program by setting a * in $path before  running.

I'm only 450 lines in... I hope you get the idea that your code can actually do bad things you do not want and this is actually extremely important.

**Comment by NovaCygni** [comment]
edge is actually spot on, just allow the forking of tools, the remerge requests into "User submitted Updates" Branch of of Manjaro, the time is now to be utilizing the community Manjaro is getting at the moment, I know in the IRC alot of us are endless repeating the same basic fixes and would much rather a way to have a more streamlined approach to fixing these issues... you can then have a "Vote for integration" section in the forum so known-working, popular and supported submitted package forks can be easily adopted into the main Manjaro project.
In short to summarize, you have a ton of extra Devs offering support, by denying this support your making the job harder for yourselves, hindering Manjaro development and thus cutting your nose simply to spite your face.

PS> At least 3/4s of the regular "Bug fixers" on your IRC havnt even a "Voice" of axx, even that little amount of recognition to people supporting Manjaro would be helpful over the situation we have atm and ease the spamming of "Help pl0x" that poor ringo suffers from bless him!

**Comment by udeved** [comment]
@edge226 

As already pointed out, feel free to send a pull request.
I do not have time to go through all these warning right now, but you do realize some of the code comes originally from upstream arch, right?
You could open the same issues on arch bug tracker for their devtools.

@NovaCygni 

Then it would be nice if ringo just opened bug reports. In fact, any user cann open issues.
>I know in the IRC alot of us are endless repeating the same basic fixes.

You do realize I am still open for pull requests? So what are you talking about with "denying access"?

**Comment by udeved** [comment]
This shellcheck of your is one typical automated tool, that is not intelligent.

~~~
In ./lib/util-iso.sh line 679:
	iso_dir="${cache_dir_iso}/${edition}/${profile}/${dist_release}/${arch}"
                 ^-- SC2154: cache_dir_iso is referenced but not assigned.
~~~

All these kind of warning can be neglected, are misleading, because these vars it complains about are assigned earlier.
Point being, this autocheck checks syntax, but does not understand the code.

**Comment by edge226** [comment]
@udeved as I stated on #212 I dont feel comfortable doing so simply because I don't know the code well enough.

Is practically copy/paste drop in test case like I put in #191 not good enough?

I stated SC2154 and SC2034 could be ignored in atom but apparently I have SC1090 in atom set. Meaning you dont see them, You can also customize the for loop I gave you in the initial post with . We know its caused by the fact we're sourcing other files for variables and if we modify the for loop to ignore them we get this

```
find ./ -type f -name '*.sh' -exec "shellcheck -x -e SC2154,SC1090" {} + > shellcheck
```
I've also sent you an updated shellcheck.txt of the above command that removes 400 of what are *real* warnings in your code.
[shellcheck.txt](https://github.com/manjaro/manjaro-tools/files/265157/shellcheck.txt)

**Comment by udeved** [comment]
The point is, you got to run this check over libs and main scripts to get reliable analysis.
Mainscripts reside in /usr/bin

**Comment by edge226** [comment]
well dont you just copy a lot of this over to /usr/bin to install it?

**Comment by udeved** [comment]
Huh?
Its bit designed like real programs.
You got libraries, and a main programs that make use of library functions.

**Comment by edge226** [comment]
@udeved I understand that... I do similar things in [my code](https://gitlab.com/orbos/dev-tools/tree/experimental).

**Comment by udeved** [comment]
I am just pointing out that the automated analysis is heavily flawed, because it only checked shared code libs.

**Comment by edge226** [comment]
That makes certain things flawed.
Not everything, many of the things that are flagged by shellcheck cause funky issues that you dont notice under X circumstance but Y it breaks.

**Comment by udeved** [comment]
Look, I have no time for this discussion.
The analysis is flawed, it complains about tons of stuff which it should not complain about if the autochecker was properly used.

**Comment by edge226** [comment]
Did you look at the newly produced file?

**Comment by udeved** [comment]
Yes, it is still partially flawed.
I don't trust automated tools more than myself.

**Comment by udeved** [comment]
May I suggest, that you do checks for each main script with its libs?
It should provide much better result.

**Comment by edge226** [comment]
So you want me to do a second find statement to include everything in bin for you? Just change the find statement to do ./bin/ instead of ./ and change it from "*.sh" to "*" to have this trigger on each file.

**Comment by udeved** [comment]
Well, you made a fuzz about almost nothing.

**Comment by edge226** [comment]
Go ask the people in #bash on freenode and see how much nothing it is. You said you were pretty much done with this as was I. Its not hard to modify the for loop I gave you, thats *why* I gave you example code. So you could modify it to fit your needs.

**Comment by udeved** [comment]
I see some missing double quotes, nothing that breaks functionality.
The rest is uncertain, and quite frankly, the checker spits out some strange complaints I do not agree with.
Anyway, all is said, 1.0.0 will fix what is to fix is this regard.

**Comment by edge226** [comment]
If you refer to this [comment](https://github.com/manjaro/manjaro-tools/issues/213#issuecomment-219296385) it shows how word splitting and globbing can negatively affect your code. Both are caused by not quoting your variables.

**Comment by udeved** [comment]
I am really tried with this discussion.
1.0.0 or send pull request.

**Comment by philmmanjaro** [comment]
@edge226: please send your pull request and I'll review it. Some of the variables are set in config files, which we load, others in our library files or bin shell scripts as most of the lib scripts are splitted to have functions we can use in several applications. Double quotes might make sense, however wont fix the **aufs** and **overlayfs** issue you face with docker, as **manjaro-tools** was/is not designed to work with it. It is always good to improve code and we are always welcome for requests and suggestions from the community. However we have also other projects we take care of and most likely have to prioritize them as they fit.
